File: Frontend/Filmer-Frontend/src/app/services/lobby.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpResponse} from '@angular/common/http';
import { Observable } from 'rxjs';

/**
 * Serwis LobbyService odpowiedzialny za obsługę lobby, w tym tworzenie, zamykanie,
 * zarządzanie uczestnikami oraz ustawieniami preferencji, a także rozpoczęcie gry.
 * @export
 * @class LobbyService
 */
@Injectable({
  providedIn: 'root'
})
export class LobbyService {
  /**
   * URL API do obsługi lobby.
   * @private
   * @type {string}
   */
  private apiUrl = 'http://localhost:8080/lobby/api';

  /**
   * Tworzy instancję LobbyService.
   * @param {HttpClient} http Serwis do wykonywania żądań HTTP.
   */
  constructor(private http: HttpClient) {}

  /**
   * Tworzy nowe lobby dla właściciela.
   * @param {number} ownerId ID właściciela lobby.
   * @returns {Observable<any>} Observable z odpowiedzią serwera.
   */
  createLobby(ownerId: number): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/${ownerId}/create`, {});
  }

  /**
   * Zamyka istniejące lobby.
   * @param {number} lobbyId ID lobby do zamknięcia.
   * @returns {Observable<void>} Observable z odpowiedzią serwera.
   */
  closeLobby(lobbyId: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${lobbyId}/close`, {});
  }

  /**
   * Dodaje użytkownika do lobby na podstawie kodu lobby.
   * @param {string} lobbyCode Kod lobby.
   * @param {number} userId ID użytkownika do dodania.
   * @returns {Observable<void>} Observable z odpowiedzią serwera.
   */
  addUserToLobby(lobbyCode: string, userId: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${lobbyCode}/addUser/${userId}`, {});
  }

  /**
   * Pobiera listę uczestników lobby na podstawie kodu lobby.
   * @param {string} lobbyCode Kod lobby.
   * @returns {Observable<any[]>} Observable zawierający listę uczestników.
   */
  getParticipants(lobbyCode: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/participants?lobbyCode=${lobbyCode}`);
  }

  /**
   * Zapisuje preferencje użytkownika dla lobby.
   * @param {string} lobbyCode Kod lobby.
   * @param {number} userId ID użytkownika.
   * @param {string} streamingPlatform Wybrana platforma streamingowa.
   * @param {string} genre Wybrany gatunek filmowy.
   * @param {string} type Wybrany typ filmu.
   * @returns {Observable<HttpResponse<any>>} Observable z odpowiedzią serwera w formacie HTTP.
   */
  savePreferences(
    lobbyCode: string,
    userId: number,
    streamingPlatform: string,
    genre: string,
    type: string
  ): Observable<HttpResponse<any>> {
    const body = { userId, streamingPlatform, genre, type };
    return this.http.post<any>(
      `${this.apiUrl}/${lobbyCode}/preferences`,
      body,
      { observe: 'response' }
    );
  }

  /**
   * Pobiera status gotowości uczestników w lobby.
   * @param {string} lobbyCode Kod lobby.
   * @returns {Observable<boolean>} Observable zawierający status gotowości.
   */
  getReadyStatus(lobbyCode: string): Observable<boolean> {
    return this.http.get<boolean>(`${this.apiUrl}/${lobbyCode}/ready-status`);
  }

  /**
   * Rozpoczyna grę w danym lobby.
   * @param {string} lobbyCode Kod lobby.
   * @returns {Observable<void>} Observable z odpowiedzią serwera.
   */
  startGame(lobbyCode: string): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${lobbyCode}/start`, {});
  }

  /**
   * Sprawdza, czy gra w danym lobby została rozpoczęta.
   * @param {string} lobbyCode Kod lobby.
   * @returns {Observable<boolean>} Observable zawierający status rozpoczęcia gry.
   */
  isGameStarted(lobbyCode: string): Observable<boolean> {
    return this.http.get<boolean>(`${this.apiUrl}/${lobbyCode}/is-started`);
  }
}

File: Backend/Filmer-Backend/src/main/java/com/filmer/filmerbackend/Services/LobbyService.java
package com.filmer.filmerbackend.Services;

import com.filmer.filmerbackend.Entities.Lobby;
import com.filmer.filmerbackend.Entities.Users;

import java.util.List;

/**
 * Interfejs serwisu do zarządzania lobby i grami.
 */
public interface LobbyService {

    /**
     * Tworzy nowe lobby z określonym właścicielem.
     *
     * @param ownerId identyfikator właściciela lobby
     * @return utworzone lobby
     */
    Lobby createLobby(int ownerId);

    /**
     * Zamyka lobby na podstawie jego identyfikatora.
     *
     * @param lobbyId identyfikator lobby
     */
    void closeLobby(int lobbyId);

    /**
     * Dodaje użytkownika do lobby na podstawie jego kodu.
     *
     * @param lobbyCode kod lobby
     * @param userId    identyfikator użytkownika
     */
    void addUserToLobby(String lobbyCode, int userId);

    /**
     * Pobiera listę uczestników lobby.
     *
     * @param lobbyCode kod lobby
     * @return lista uczestników lobby
     */
    List<Users> getParticipants(String lobbyCode);

    /**
     * Sprawdza, czy wszyscy użytkownicy w lobby są gotowi.
     *
     * @param lobbyId identyfikator lobby
     * @return true, jeśli wszyscy użytkownicy są gotowi, false w przeciwnym razie
     */
    boolean areAllUsersReady(int lobbyId);

    /**
     * Zapisuje preferencje użytkownika dotyczące platformy streamingowej, gatunku i typu filmu.
     *
     * @param lobbyCode kod lobby
     * @param userId    identyfikator użytkownika
     * @param streamingPlatform preferowana platforma streamingowa
     * @param genre     preferowany gatunek
     * @param type      preferowany typ filmu
     */
    void saveUserPreferences(String lobbyCode, int userId, String streamingPlatform, String genre, String type);

    /**
     * Pobiera lobby na podstawie jego kodu.
     *
     * @param lobbyCode kod lobby
     * @return obiekt lobby
     */
    Lobby getLobbyByCode(String lobbyCode);

    /**
     * Rozpoczyna grę w określonym lobby.
     *
     * @param lobbyId identyfikator lobby
     */
    void startGame(int lobbyId);

    /**
     * Sprawdza, czy głosowanie w lobby zostało zakończone.
     *
     * @param lobbyCode kod lobby
     * @return true, jeśli głosowanie zostało zakończone, false w przeciwnym razie
     */
    boolean checkVotingCompletion(String lobbyCode);

    /**
     * Kończy głosowanie dla użytkownika w lobby.
     *
     * @param lobbyCode kod lobby
     * @param userId    identyfikator użytkownika
     */
    void finishVoting(String lobbyCode, int userId);
}

